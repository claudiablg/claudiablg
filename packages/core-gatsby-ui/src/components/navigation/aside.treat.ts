import { Theme } from '@newrade/core-react-ui';
import { style } from 'treat';

export const styles = {
  wrapper: style(({ cssTheme, theme }: Theme) => ({
    visibility: 'hidden',
    position: 'fixed',
    top: 80,
    right: 40,
    width: cssTheme.layout.asideWidth,
    padding: '1em',

    gap: cssTheme.sizing.var.x2,

    backgroundColor: 'transparent',

    '@media': {
      [`screen and (min-width: ${theme.layout.breakpoints.desktopMedium}px)`]: {
        visibility: 'visible',
      },
    },
  })),
  link: style(({ cssTheme, theme }: Theme) => ({
    color: 'inherit',
    textDecoration: 'none',
    fontWeight: 400,
  })),
  linkSelected: style(({ cssTheme, theme }: Theme) => ({
    fontWeight: 600,
  })),
};
