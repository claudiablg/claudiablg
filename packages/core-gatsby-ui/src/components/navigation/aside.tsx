import { Label, Stack, Box, useTreatTheme } from '@newrade/core-react-ui';
import { WindowLocation } from '@reach/router';
import { kebab } from 'case';
import React from 'react';
import { useStyles } from 'react-treat';
import { useLocation } from '../../hooks/use-reach-router-location';
import * as styleRefs from './aside.treat';
import { formatAnchorId } from '@newrade/core-react-ui';

type AsideItem<TState = any> = {
  value?: string | null;
  depth?: number | null;
} | null;

type Props = {
  items?: AsideItem[] | null;
  location?: WindowLocation<any>;
};

export const Aside: React.FC<Props> = (props) => {
  const { styles } = useStyles(styleRefs);
  const { theme, cssTheme } = useTreatTheme();

  const id = props?.location?.hash;
  const currentId = useScrollSpy(props.items);

  return (
    <Box as={'aside'} className={styles.wrapper}>
      <Stack as={'nav'} className={styles.wrapper}>
        {props.items?.filter(filterItemDepthPredicate).map((item) => {
          const href = `#${formatAnchorId(item?.value)}`;
          const selected = formatAnchorId(item?.value) === currentId;
          return (
            <a href={href} className={styles.link} key={formatAnchorId(item?.value)}>
              <Label
                id={`aside-link-${item?.value}`}
                className={`${styles.link} ${selected ? styles.linkSelected : ''}`}
              >
                {item?.value}
              </Label>
            </a>
          );
        })}
      </Stack>
    </Box>
  );
};

function filterItemDepthPredicate(item: AsideItem) {
  return item !== undefined && item !== null && item.depth && item.depth > 1;
}

function useScrollSpy(items: Props['items']) {
  const [currentId, setCurrentId] = React.useState<string | null>(null);
  const location = useLocation();

  // React.useEffect(() => setCurrentId(null), [location?.hash]);

  React.useEffect(() => {
    if (!(items && items.length && items.filter(filterItemDepthPredicate).length)) {
      setCurrentId(null);
      return;
    }

    const elements = document.querySelectorAll<HTMLElement>(
      items
        .filter(filterItemDepthPredicate)
        .map((item) => `[id="${formatAnchorId(item?.value)}"]`)
        .join(',')
    );
    const elementsArray = Array.from(elements);

    const handleScroll = () => {
      elementsArray.forEach((element) => {
        if (element.offsetTop <= window.scrollY + 40) {
          console.log(element.offsetTop);
          console.log(window.scrollY);
          setCurrentId(element.id);
        }
      });
    };
    window.addEventListener('scroll', handleScroll);
    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, [items]);

  return currentId;
}
