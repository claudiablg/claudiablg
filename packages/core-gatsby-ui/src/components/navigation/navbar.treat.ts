import { Theme } from '@newrade/core-react-ui';
import { style } from 'treat';

export const styles = {
  wrapper: style(({ cssTheme, theme }: Theme) => ({
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    zIndex: 100,
    height: `56px`,

    backgroundColor: 'white',
    boxShadow: '0 1px 2px rgba(33,33,33,0.15)',
  })),

  container: style(({ cssTheme, theme }: Theme) => ({
    maxWidth: '1200px',
  })),

  link: style(({ cssTheme, theme }: Theme) => ({
    width: 'fit-content',
  })),

  menuWrapper: style(({ cssTheme, theme }: Theme) => ({
    width: 'fit-content',
  })),
};
