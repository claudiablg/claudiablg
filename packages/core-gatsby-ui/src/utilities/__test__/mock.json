{
  "component": "sidebar",
  "links": [],
  "name": "test",
  "path": "/",
  "subNavigation": [
    {
      "component": "menu",
      "label": "Design System",
      "links": [],
      "name": "index",
      "path": "/design-system/",
      "subNavigation": [
        {
          "component": "menu",
          "label": "Components",
          "links": [],
          "name": "index",
          "path": "/design-system/components/",
          "subNavigation": [
            {
              "component": "link",
              "label": "Components",
              "link": {
                "label": "Overview",
                "name": "index",
                "page": {
                  "name": "index",
                  "slug": "/design-system/components/",
                  "title": "Overview"
                },
                "type": "internalPage"
              },
              "links": [],
              "name": "index",
              "subNavigation": []
            },
            {
              "component": "link",
              "label": "Buttons",
              "link": {
                "label": "Buttons",
                "name": "buttons",
                "page": {
                  "name": "buttons",
                  "slug": "/design-system/components/buttons/",
                  "title": "Buttons"
                },
                "type": "internalPage"
              },
              "links": [],
              "name": "buttons",
              "path": "/design-system/components/buttons/",
              "subNavigation": []
            }
          ]
        },
        {
          "component": "link",
          "label": "Design System",
          "link": {
            "label": "Overview",
            "name": "index",
            "page": {
              "name": "index",
              "slug": "/design-system/",
              "title": "Overview"
            },
            "type": "internalPage"
          },
          "links": [],
          "name": "index",
          "subNavigation": []
        }
      ]
    },
    {
      "component": "menu",
      "label": "Docs",
      "links": [],
      "name": "index",
      "path": "/docs/",
      "subNavigation": [
        {
          "component": "link",
          "label": "Docs",
          "link": {
            "label": "Overview",
            "name": "index",
            "page": {
              "name": "index",
              "slug": "/docs/",
              "title": "Overview"
            },
            "type": "internalPage"
          },
          "links": [],
          "name": "index",
          "subNavigation": []
        }
      ]
    }
  ]
}
