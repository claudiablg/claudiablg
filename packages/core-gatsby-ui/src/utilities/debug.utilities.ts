import debug from 'debug';

/**
 * CLI debug namespace
 */
export const NS = 'nr:core-gatsby-ui';

export const debugInstance = debug;

export const log = debugInstance(`${NS}`);
