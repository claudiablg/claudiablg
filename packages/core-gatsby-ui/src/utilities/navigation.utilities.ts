import { Navigation } from '../models/navigation.model';
import { NavItem } from '../models/nav-item.model';
import { GatsbyCommonPageContext } from '@newrade/core-gatsby-config';
import { title, lower, kebab } from 'case';

/**
 * Simplified type for a Page Node
 * uses the same context that is used to create all pages.
 */
export type PageNode = {
  path: string;
  context?: GatsbyCommonPageContext | null;
};

/**
 * Returns a Navigation object from page nodes.
 * Accepts arrays of values to sort the directories and items.
 */
export function getNavigationFromPageNodes({
  name,
  pageNodes,
  sortOrderDirectories,
  sortOrderItems,
  ignoredItems,
  formatName,
  formatDisplayName,
}: {
  name: string;
  pageNodes?: PageNode[];
  sortOrderItems?: string[];
  sortOrderDirectories?: string[];
  ignoredItems?: string[];
  formatName?: (name?: string | null) => string;
  formatDisplayName?: (name?: string | null) => string;
}): Navigation {
  if (!pageNodes?.length) {
    return {
      name,
      items: [],
    };
  }

  const normalizedSortOrderItems = sortOrderItems?.map((item) => normalizeName(item));
  const normalizedSortOrderDirectories = sortOrderDirectories?.map((item) => normalizeName(item));
  const normalizedIgnoredItems = ignoredItems?.map((item) => normalizeName(item));

  // find the dirNames
  const dirNames = pageNodes
    // remove 404 pages
    .filter((node) => !/404/gi.test(node.context?.name || ''))
    // remove ignored items
    .filter((node) => !normalizedIgnoredItems?.find((item) => item === normalizeName(node.context?.name)))
    .map((node) => node?.context?.dirName);

  // for item at the root the dir name is ''
  const navigation = [...new Set(dirNames)].reduce(
    (previous, current) => {
      // for each dir name, transform the nodes and place them in .items
      const currentDirName = current;

      const dirNameNodes = pageNodes.filter((node) => node?.context?.dirName === currentDirName);
      if (!dirNameNodes) {
        return previous;
      }

      previous.items = [
        ...previous.items,
        // create a new item
        {
          name: currentDirName ? currentDirName : '',
          displayName: formatDisplayName ? formatDisplayName(currentDirName) : defaultFormatDisplayName(currentDirName),
          path: currentDirName ? currentDirName : '',
          // for each child nodes, creat a nav item
          items: dirNameNodes
            .map((node) => getNavItemShallow(node))
            // apply sorting on items, if specified
            .sort((a, b) => {
              if (!normalizedSortOrderItems) {
                return 0;
              }

              const indexA = normalizedSortOrderItems.indexOf(normalizeName(a.name));
              const indexB = normalizedSortOrderItems.indexOf(normalizeName(b.name));

              if (indexA === indexB) {
                return 0;
              }

              if (indexA === -1) {
                return 1;
              }

              if (indexB === -1) {
                return -1;
              }

              return indexA > indexB ? 1 : -1;
            }),
        },
        // apply sorting on directories, if specified
      ].sort((a, b) => {
        if (!normalizedSortOrderDirectories) {
          return 0;
        }

        const indexA = normalizedSortOrderDirectories.indexOf(normalizeName(a.name));
        const indexB = normalizedSortOrderDirectories.indexOf(normalizeName(b.name));

        if (indexA === indexB) {
          return 0;
        }

        if (indexA === -1) {
          return 1;
        }

        if (indexB === -1) {
          return -1;
        }

        return indexA > indexB ? 1 : -1;
      });

      return previous;
    },
    {
      name,
      items: [],
    } as Navigation
  );

  return navigation;

  function getNavItemShallow(node: PageNode): NavItem {
    return {
      name: formatName ? formatName(node?.context?.name) : defaultFormatName(node?.context?.name),
      displayName: formatDisplayName
        ? formatDisplayName(node?.context?.name || node?.context?.displayName)
        : defaultFormatDisplayName(node?.context?.name || node?.context?.displayName),
      // leave the node path untouched
      path: node.path,
    };
  }
}

function normalizeName(name?: string | null) {
  return kebab(name || '');
}

function defaultFormatDisplayName(name?: string | null): string {
  if (!name) {
    return '';
  }
  return title(defaultFormatName(name));
}

function defaultFormatName(name?: string | null): string {
  if (!name) {
    return '';
  }
  return name
    .replace('.page', '')
    .replace('design-system-', '')
    .replace(/\s-\s$/, '');
}
