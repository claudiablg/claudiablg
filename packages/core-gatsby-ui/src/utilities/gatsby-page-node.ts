import {
  GatsbyCommonPageContext,
  GatsbyMarkdownFilePageContext,
} from '@newrade/core-gatsb-config/config';
import { PartialOrNull } from '@newrade/core-website-api/utilities';

/**
 * Simplified type for a Page Node
 * uses the same context that is used to create all pages.
 *
 * It assumes that all values can be null or undefined (since null is the value
 * returned in graphql when something is not defined)
 */
export type GatsbyPageNode = PartialOrNull<{
  id: string;
  path: string;
  context?: GatsbyCommonPageContext | GatsbyMarkdownFilePageContext | null;
}>;
