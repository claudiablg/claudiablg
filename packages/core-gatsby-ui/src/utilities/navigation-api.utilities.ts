import { SITE_LANGUAGES } from '@newrade/core-common';
import { NavComponent, NavigationAPI } from '@newrade/core-website-api';
import { PartialOrNull } from '@newrade/core-website-api/src/utilities';
import { GatsbyPageNode } from './gatsby-page-node';

export type GetNavigationAPIOptions = {
  /**
   * Page nodes to transform
   */
  pageNodes?: PartialOrNull<GatsbyPageNode>[];
  /**
   * Name of the navigation object
   * @example 'footer'
   */
  navigationName: string;
  /**
   * Navigation component
   */
  navigationComponent: NavComponent;
  /**
   * Include only the listed locales
   */
  includeLocales?: SITE_LANGUAGES[];
  /**
   * Sort order for leafs navigation items
   */
  sortOrderItems?: string[];
  /**
   * Sort order for directories
   */
  sortOrderDirectories?: string[];
  /**
   * Excluded paths
   * @example '/404'
   */
  excludePaths?: string[];
  /**
   * Words to transform to UPPERCASE
   */
  uppercaseWords?: string[];
  /**
   * Translate function (is executed before formatLabel)
   */
  translate?: (key?: string | null) => string;
  /**
   * Format function applied to label or name if label does not exist
   */
  formatLabel?: (name?: string | null) => string;
};

const defaultUppercaseWords = [
  'wsl',
  'ui',
  'ux',
  'seo',
  'ssh',
  'css',
  'api',
  'ci',
  'vm',
  'cms',
  'pr',
  'cli',
];

/**
 * Returns a Navigation object from page nodes.
 * Accepts arrays of values to sort the directories and items.
 */
export function getNavigationAPIFromPageNodes(options: GetNavigationAPIOptions): NavigationAPI {
  if (!options.pageNodes?.length) {
    return {
      id: options.navigationName,
      name: options.navigationName,
      label: '',
      component: NavComponent.sidebar,
      links: [],
      subNavigation: [],
    };
  }

  // build tree navigation structure

  // format name and label

  // create links

  const uppercaseWords = options.uppercaseWords || defaultUppercaseWords;

  return {
    id: options.navigationName,
    name: options.navigationName,
    label: '',
    component: NavComponent.sidebar,
    links: [],
    subNavigation: [],
    tags: [],
  };
}
