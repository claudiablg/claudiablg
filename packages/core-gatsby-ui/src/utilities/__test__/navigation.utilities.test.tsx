import { Navigation } from '../../navigation/navigation.model';
import { getNavigationFromPageNodes, getPageDirFromPath } from '../navigation.utilities';
import { navigationMock } from './navigation.mock';

describe(`navigation utilities`, () => {
  describe(`${getPageDirFromPath.name}`, () => {
    it('should extract the page dir name from a path with no locale', () => {
      expect(getPageDirFromPath('/design-system/page-name.tsx')).toBe('');
      expect(getPageDirFromPath('dir-name/page-name.tsx')).toBe('dir-name');
      expect(getPageDirFromPath('dir-name/page-name.md')).toBe('dir-name');
      expect(getPageDirFromPath('dir-name/page-name.mdx')).toBe('dir-name');
      expect(getPageDirFromPath('/page-name.tsx')).toBe('');
      expect(getPageDirFromPath('/index.tsx')).toBe('');
      expect(getPageDirFromPath('index.tsx')).toBe('');
      expect(getPageDirFromPath('page-name.tsx')).toBe('');
      expect(getPageDirFromPath('/dir/page-name.tsx')).toBe('dir');
    });

    it('should extract the page dir name from a path with locales', () => {
      expect(getPageDirFromPath('/fr/design-system/page-name.tsx')).toBe('');
      expect(getPageDirFromPath('/fr/dir-name/page-name.tsx')).toBe('dir-name');
      expect(getPageDirFromPath('/fr/dir-name/page-name.md')).toBe('dir-name');
      expect(getPageDirFromPath('/fr/dir-name/page-name.mdx')).toBe('dir-name');
      expect(getPageDirFromPath('/fr/')).toBe('');
      expect(getPageDirFromPath('/en/')).toBe('');
      expect(getPageDirFromPath('/fr/page-name.tsx')).toBe('');
      expect(getPageDirFromPath('/fr/page-name.tsx')).toBe('');
      expect(getPageDirFromPath('/fr/dir/page-name.tsx')).toBe('dir');
    });
  });

  describe(`utility ${getNavigationFromPageNodes.name}`, () => {
    const emptyNavigation: Navigation = {
      name: 'navigation test',
      items: [],
    };

    it('should return an empty Navigation object from Gatsby page nodes', () => {
      const nav = getNavigationFromPageNodes({
        name: 'navigation test',
        pageNodes: [],
      });

      expect(nav).toEqual(emptyNavigation);
    });

    it('should return a valid Navigation object from Gatsby page nodes', () => {
      const nav = getNavigationFromPageNodes(navigationMock);

      const expectedNav: Navigation = {
        name: 'navigation test',
        items: [
          {
            name: 'path-a',
            displayName: 'Path A',
            path: 'path-a',
            items: [
              {
                name: 'page-2',
                displayName: 'Page 2',
                path: '/design-system/path-a/page-2',
              },
              {
                name: 'page-1',
                displayName: 'Page 1',
                path: '/design-system/path-a/page-1',
              },
            ],
          },
          {
            name: 'Home',
            displayName: '',
            path: '',
            items: [
              {
                name: 'index',
                displayName: 'Home',
                path: '/',
              },
              {
                name: 'home',
                displayName: 'Home',
                path: '/design-system/',
              },
            ],
          },
          {
            name: 'path-b',
            displayName: 'Path B',
            path: 'path-b',
            items: [
              {
                name: 'page-1',
                displayName: 'Page 1',
                path: '/design-system/path-b/page-1',
              },
            ],
          },
        ],
      };

      expect(nav).toEqual(expectedNav);
    });
  });
});
