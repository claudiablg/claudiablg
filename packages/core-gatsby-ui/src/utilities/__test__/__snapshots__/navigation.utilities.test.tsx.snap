// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation utilities utility getNavigationFromPageNodes should return a valid Navigation object from Gatsby page nodes 1`] = `
Object {
  "items": Array [
    Object {
      "displayName": "Path A",
      "items": Array [
        Object {
          "displayName": "Overview",
          "frontmatter": undefined,
          "name": "index",
          "path": "/design-system/path-a/",
        },
        Object {
          "displayName": "Page 2",
          "frontmatter": undefined,
          "name": "page-2",
          "path": "/design-system/path-a/page-2",
        },
        Object {
          "displayName": "Page 1",
          "frontmatter": undefined,
          "name": "page-1",
          "path": "/design-system/path-a/page-1",
        },
      ],
      "name": "path-a",
      "path": "path-a",
    },
    Object {
      "displayName": "",
      "items": Array [
        Object {
          "displayName": "Overview",
          "frontmatter": undefined,
          "name": "index",
          "path": "/",
        },
        Object {
          "displayName": "Overview",
          "frontmatter": undefined,
          "name": "index",
          "path": "/design-system/",
        },
      ],
      "name": "Home",
      "path": "",
    },
    Object {
      "displayName": "Path B",
      "items": Array [
        Object {
          "displayName": "Page 1",
          "frontmatter": undefined,
          "name": "page-1",
          "path": "/design-system/path-b/page-1",
        },
      ],
      "name": "path-b",
      "path": "path-b",
    },
  ],
  "name": "navigation test",
}
`;
